src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HashProviderDispenser.OpenSsl.cs

```diff
@@ -4,53 +4,42 @@
 using System;
 using System.Diagnostics;
 using System.Runtime.InteropServices;
-using System.Security.Cryptography;
 using Internal.Cryptography;
 using Microsoft.Win32.SafeHandles;
 
 namespace System.Security.Cryptography
 {
     internal static partial class HashProviderDispenser
     {
-        internal static HashProvider CreateHashProvider(string hashAlgorithmId)
+        internal static bool KmacSupported(string algorithmId)
         {
-            return new EvpHashProvider(hashAlgorithmId);
-        }
-
-        internal static HashProvider CreateMacProvider(string hashAlgorithmId, ReadOnlySpan<byte> key)
-        {
-            return new HmacHashProvider(hashAlgorithmId, key);
-        }
-
-        internal static bool HashSupported(string hashAlgorithmId)
-        {
-            return Interop.Crypto.HashAlgorithmSupported(hashAlgorithmId);
+            return algorithmId switch
+            {
+                HashAlgorithmNames.KMAC128 => Interop.Crypto.EvpMacAlgs.Kmac128 is not null,
+                HashAlgorithmNames.KMAC256 => Interop.Crypto.EvpMacAlgs.Kmac256 is not null,
+                _ => false,
+            };
         }
 
-        internal static bool MacSupported(string hashAlgorithmId) => HashSupported(hashAlgorithmId);
-
-        internal static class OneShotHashProvider
+        internal static partial class OneShotHashProvider
         {
-            public static int MacData(
-                string hashAlgorithmId,
+            public static void KmacData(
+                string algorithmId,
                 ReadOnlySpan<byte> key,
                 ReadOnlySpan<byte> source,
-                Span<byte> destination)
+                Span<byte> destination,
+                ReadOnlySpan<byte> customizationString,
+                bool xof)
             {
-                IntPtr evpType = Interop.Crypto.HashAlgorithmToEvp(hashAlgorithmId);
-                Debug.Assert(evpType != IntPtr.Zero);
-
-                int hashSize = Interop.Crypto.EvpMdSize(evpType);
-
-                if (hashSize <= 0 || destination.Length < hashSize)
+                SafeEvpMacHandle? macHandle = algorithmId switch
                 {
-                    Debug.Fail("Destination length or hash size not valid.");
-                    throw new CryptographicException();
-                }
+                    HashAlgorithmNames.KMAC128 => Interop.Crypto.EvpMacAlgs.Kmac128,
+                    HashAlgorithmNames.KMAC256 => Interop.Crypto.EvpMacAlgs.Kmac256,
```

Comment:
Yeah. Fixing rooting for AOT deep in our abstractions is bigger than KMAC. I'm not sure its worth it either, and it would force pushing abstractions all the way up to the public API surface.